In [128]:
import pandas as pd
import numpy as np
import json

Importing the data as a dataframe.

In [129]:
dataset = pd.read_csv("data/Data_Lv3_UMDOTS_Escooters.csv")

Lets take a look at a few data point

In [130]:
dataset.head()

Unnamed: 0  Unnamed: 0.1                    CREATED  \
0           0             0  2020-10-03 03:32:15+00:00   
1           1             1  2020-10-03 03:07:39+00:00   
2           2             2  2020-10-03 03:17:05+00:00   
3           3             3  2020-10-03 02:24:58+00:00   
4           4             4  2020-10-03 03:19:34+00:00   

                       START                        END  DISTANCE  RIDE_ID  \
0  2020-10-03 03:32:15+00:00  2020-10-03 03:32:46+00:00     0.008  3135665   
1  2020-10-03 03:07:39+00:00  2020-10-03 03:37:06+00:00     1.062  3135395   
2  2020-10-03 03:17:05+00:00  2020-10-03 03:24:24+00:00     0.847  3135503   
3  2020-10-03 02:24:58+00:00  2020-10-03 03:11:56+00:00     3.447  3134955   
4  2020-10-03 03:19:34+00:00  2020-10-03 03:22:45+00:00     0.386  3135529   

   CUSTOMER_ID  MINUTES VEHICLE_TYPE  \
0       193964        1      Scooter   
1       542947       30      Scooter   
2       517321        8      Scooter   
3       578051       47      Scooter   
4       552305        4      Scooter   

                                                PATH  \
0  {\n  "coordinates": [\n    [\n      -76.934163...   
1  {\n  "coordinates": [\n    [\n      -76.938180...   
2  {\n  "coordinates": [\n    [\n      -76.937599...   
3  {\n  "coordinates": [\n    [\n      -76.943553...   
4  {\n  "coordinates": [\n    [\n      -76.938115...   

                                          TIMESTAMPS  START LONG  START LAT  \
0  [\n  "2020-10-03T03:32:14Z",\n  "2020-10-03T03...  -76.934163  38.990656   
1  [\n  "2020-10-03T03:07:46Z",\n  "2020-10-03T03...  -76.938181  38.982087   
2  [\n  "2020-10-03T03:17:12Z",\n  "2020-10-03T03...  -76.937600  38.979613   
3  [\n  "2020-10-03T02:25:04Z",\n  "2020-10-03T02...  -76.943553  38.982470   
4  [\n  "2020-10-03T03:19:40Z",\n  "2020-10-03T03...  -76.938116  38.982084   

    END LONG    END LAT  
0 -76.934285  38.990739  
1 -76.949165  38.992849  
2 -76.933335  38.992676  
3 -76.943546  38.982476  
4 -76.943257  38.986101

In [131]:
dataset.shape

(40325, 16)

The dataset has 40325 rows and 16 columns

In [132]:
# All the data types do not match lets fix them

# Convert as Timestamp data
dataset["CREATED"] = pd.to_datetime(dataset["CREATED"])
# Convert as Timestamp data
dataset["START"] = pd.to_datetime(dataset["START"])
# Convert as Timestamp data
dataset["END"] = pd.to_datetime(dataset["END"])
# Convert timestamp as list
dataset["TIMESTAMPS"] = dataset["TIMESTAMPS"].apply(eval)

In [133]:
dataset.dtypes

Unnamed: 0                    int64
Unnamed: 0.1                  int64
CREATED         datetime64[ns, UTC]
START           datetime64[ns, UTC]
END             datetime64[ns, UTC]
DISTANCE                    float64
RIDE_ID                       int64
CUSTOMER_ID                   int64
MINUTES                       int64
VEHICLE_TYPE                 object
PATH                         object
TIMESTAMPS                   object
START LONG                  float64
START LAT                   float64
END LONG                    float64
END LAT                     float64
dtype: object

In [134]:
dataset.isna().any()

Unnamed: 0      False
Unnamed: 0.1    False
CREATED         False
START           False
END             False
DISTANCE         True
RIDE_ID         False
CUSTOMER_ID     False
MINUTES         False
VEHICLE_TYPE    False
PATH            False
TIMESTAMPS      False
START LONG      False
START LAT       False
END LONG        False
END LAT         False
dtype: bool

The distance column has missing values

In [135]:
dataset["DISTANCE"].isna().sum()
#dataset["DISTANCE"].dropna()

2273

Total of 2273 missing datapoint for distance. No sure how to deal with this yet.

Next we separate all ride by years: October 2020 and October 2019

In [139]:
dataset["Year"] = 2010

In [123]:
# Get the year of each data entry
for index, item in dataset.iterrows():
    dataset["Year"]["index"] = dataset["CREATED"][index].year

<ipython-input-123-502c50988992>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataset["Year"]["index"] = dataset["CREATED"][index].year


KeyboardInterrupt: 

In [97]:
ride_count_year = dataset.groupby("Year").size().reset_index(name='ride_count')
ride_count_year

Year  ride_count
0  2019       18382
1  2020       21943